app/src/cc/arduino/packages/formatter/AStyle.java

```diff
@@ -76,15 +80,64 @@ public void init(Editor editor) {
 
   @Override
   public void run() {
-    String originalText = editor.getCurrentTab().getText();
-    String formattedText = aStyleInterface.AStyleMain(originalText, formatterConfiguration);
 
-    if (formattedText.equals(originalText)) {
-      editor.statusNotice(tr("No changes necessary for Auto Format."));
-      return;
-    }
+    SketchTextArea textArea = editor.getCurrentTab().getTextArea();
+
+    String originalText = textArea.getSelectedText();
+
+    // If no selection use all file
+    if (originalText == null || originalText.isEmpty()) {
+
+      String formattedText = aStyleInterface.AStyleMain(textArea.getText(), formatterConfiguration);
+      editor.getCurrentTab().setText(formattedText);
+
+    } else {
+      try {
+
+        // apply indentation control keywords.
+        String FORMAT_ON = "\n// *INDENT-ON* DYN\n";
+        String FORMAT_OFF = "\n// *INDENT-OFF* DYN\n";
 
-    editor.getCurrentTab().setText(formattedText);
+        RSyntaxDocument content = (RSyntaxDocument) textArea.getDocument();
+
+        textArea.beginAtomicEdit();
+
+        int selStart = editor.getCurrentTab().getSelectionStart();
+        int selEnd = editor.getCurrentTab().getSelectionStop();
+        int lineStart = textArea.getLineOfOffset(selStart);
+        int lineEnd = textArea.getLineOfOffset(selEnd);
+
+        // Calculate offsets from begin and end of each line.
+        int fristLineOffset = textArea.getLineStartOffset(lineStart);
+        int lastLineOffset = textArea.getLineEndOffset(lineEnd);
+
+        // inserts change the length, use this to calculate new positios.
+        int offLength = FORMAT_OFF.length();
+        int onLength = FORMAT_ON.length();
+
+        content.insertString(0, FORMAT_OFF, null);
+        content.insertString(fristLineOffset + offLength, FORMAT_ON, null);
+        content.insertString(lastLineOffset + offLength + onLength, FORMAT_OFF,null);
+        originalText = content.getText(0, content.getLength());
+
+        String formattedText = aStyleInterface.AStyleMain(originalText, formatterConfiguration);
+
+        // Remove format tags
+        formattedText = formattedText.replaceAll(Pattern.quote(FORMAT_OFF), "");
+        formattedText = formattedText.replaceAll(Pattern.quote(FORMAT_ON), "");
+
+        textArea.setText(formattedText);
+        textArea.select(selStart, selStart);
```

Comment:
AFAICS, this clears the selection and puts the cursor at the start? Might be good to keep the selection (requires recalculating the end of the selection, but that is probably easier if you solve my previous comment too).

Reply:
This will place the cursor at the beginning of the SELECTION. I would have to recalculate the positioning of the new format, to keep things simple and straightforward I decided not to do it.

Reply:
No negative point for usability